uint8[] buffer
uint32 len
---
uint32 totalWritten
